uarray/_uarray_dispatch.cxx

```diff
@@ -1816,7 +1816,6 @@ extern "C" MODULE_EXPORT PyObject * PyInit__uarray(void) {
       PyExc_NotImplementedError, nullptr));
   if (!BackendNotImplementedError)
     return nullptr;
-  Py_INCREF(BackendNotImplementedError.get());
```

Comment:
I don't see why an extra `INCREF` would cause segfaults. Would you link to a reproducer in an issue, add a test, and mark the PR as fixing the issue?

Reply:
I will try to reproduce. The main problem here is that an additional reference delays removal of [BackendNotImplementedError  ](https://github.com/Quansight-Labs/uarray/blob/a5c3be24/uarray/_uarray_dispatch.cxx#L131) till a destructor call that will happen when dynamic loader will unload a dynamic library and the object is not freed in `globals_free`. In case of an embedded interpreter calling `Py_XDECREF(obj_);` after a `Py_FinalizeEx` call causes a fault.

Reply:
How about `importlib.reload(uarray)`?

Reply:
Sorry I can not reproduce the issue in a test environment. I've tried to reload  the module similar to 
```
libdl = ctypes.CDLL("libdl.so")
libdl.dlclose.argtypes = [ctypes.c_void_p]
so = ctypes.PyDLL("./uarray/_uarray.cpython-38-x86_64-linux-gnu.so")
so.PyInit__uarray.argtypes = []
so.PyInit__uarray.restype = ctypes.py_object
m = so.PyInit__uarray()
err = m.BackendNotImplementedError
del m
gc.collect(2)
libdl.dlclose(so._handle)
```
in a separate Pythnon process. It works fine and I can force so module load and reload but there is no way to call  ~py_ref for  `static py_ref BackendNotImplementedError`. So all tests are similar to a reference counter check.

Calling `.release()` has another effect as removing INCREF: it sets a nullptr to the error object  and tests show multiple assertion errors like
```
E       SystemError: <built-in function determine_backend> returned NULL without setting an error
```

Reply:
I would just wait for @peterbell10's approval and CI before merging.

Reply:
CI is red, but locally tests work for me. 
I will try to dig deeper why removing `Py_INCREF` line fixes crashes at exit when the interpreter is already finalized.